sys/xtimer/xtimer.c

```diff
@@ -237,34 +244,66 @@ int _xtimer_msg_receive_timeout(msg_t *msg, uint32_t timeout_ticks)
     return _msg_wait(msg, &tmsg, &t);
 }
 
+static int _mutex_remove_thread_from_waiting_queue(mutex_t *mutex, thread_t *thread)
+{
+    unsigned irqstate = irq_disable();
+    assert(mutex != NULL && thread != NULL);
+    DEBUG("_mutex_remove_thread_from_waiting_queue(): param thread: %"PRIkernel_pid"\n", thread->pid);
+
+    if (mutex->queue.next != MUTEX_LOCKED && mutex->queue.next != NULL) {
+        list_node_t *node = list_remove(&mutex->queue, (list_node_t *)&thread->rq_entry);
+        /* if thread was removed from the list */
+        if (node != NULL) {
+            DEBUG("thread: %"PRIkernel_pid" removed from list\n", (container_of((clist_node_t *)node, thread_t, rq_entry)->pid));
+            if (mutex->queue.next == NULL) {
+                mutex->queue.next = MUTEX_LOCKED;
+            }
+            sched_set_status(thread, STATUS_PENDING);
+            irq_restore(irqstate);
+            thread_yield_higher();
```

Comment:
It looks like it is `sched_switch` that may need to be used here instead to be working everywhere.
But from the implementation, `thread_yield_higher` in `cpu/atmega_common/thread_arch.c`, does the check for `is_in_irq`…

I will do a dedicated test for this.

Reply:
https://github.com/RIOT-OS/RIOT/pull/11759

Reply:
Implementation in the CPU is wrong, this code is correct https://github.com/RIOT-OS/RIOT/pull/11759#issuecomment-506760423

Reply:
I will change it to `sched_switch` for now until #11759 is solved/discussed/closed.

Reply:
done in sys/xtimer/xtimer.c: _mutex_timeout() clean ups #11807